Automattic/wp-calypso
# Plans 2023: Rewire plan availability-for-purchase checks

## Proposed Changes

- This is a general rewrite and rewire of how we compute the availability of a plan for purchasing. See discussion here: https://github.com/Automattic/wp-calypso/pull/83940#discussion_r1424505087 These parts were recently refactored a bit, which led to concerns being raised by how the hook, underlying selector, and usage in pricing meta come together. It was fairly ambiguous, we were mixing "upgrade" with "purchase", hard to tell what the hook was really doing, and the same logic was being redone almost in reverse order in `usePricingMetaForGridPlans`. These issues are what the changes here aim to address.
- It also uncouples the dependency on a state selector in `usePricingMetaForGridPlans` (this hook is being ported into the package next to compile the pricing of the grid plans off the data-store hooks).

## Testing Instructions



* Everything about pricing and labels shown for the plans
* Compare with production:
    * `/plans/[site]` with a site and a current plan 
    * Prices in `/start/plans` are correct
    * Currency discounts are shown with the respective label (try INR or other currencies)
    * Prorated prices in `/plans[site]` are correct (shown with respective "upgrade credits" label)
    * Try with a site that you are not the owner (the non-discounted/prorated prices should be shown)
    * Anything else pricing-related that comes to mind

## Pre-merge Checklist
